lib/cereal/patterns/chain.go

```diff
@@ -0,0 +1,333 @@
+package patterns
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+
+	"github.com/chef/automate/lib/cereal"
+)
+
+var ErrIncorrectParameters = errors.New("Incorrect number of parameters given")
+var ErrTaskWorkflowInvalid = errors.New("Could not determine workflow for task")
+
+type ChainWorkflowParams struct {
+	WorkflowParams []json.RawMessage
+}
+
+type ChainWorkflowPayload struct {
+	Canceled bool
+	State    []WorkflowState
+}
+
+func (p *ChainWorkflowPayload) IsValid() bool {
+	return len(p.State) > 0
+}
+
+func (p *ChainWorkflowPayload) Finished() bool {
+	for _, state := range p.State {
+		if !state.IsFinished {
+			return false
+		}
+	}
+	return true
+}
+
+type ChainWorkflowExecutor struct {
+	executors []cereal.WorkflowExecutor
+}
+
+type ChainWorkflowTaskParam struct {
+	XXX_ChainWorkflowIdx int64 `json:"__idx"`
+}
+
+func NewChainWorkflowExecutor(executors ...cereal.WorkflowExecutor) (*ChainWorkflowExecutor, error) {
+	if len(executors) <= 0 {
+		return nil, errors.New("Can't have zero")
+	}
+	return &ChainWorkflowExecutor{executors: executors}, nil
+}
+
+func ToChainWorkflowParameters(parameters []interface{}) (ChainWorkflowParams, error) {
+	params := ChainWorkflowParams{}
+	if len(parameters) == 0 {
+		return params, errors.New("chain workflow parameters empty")
+	}
+	params.WorkflowParams = make([]json.RawMessage, len(parameters))
+	for idx, subWorkflowParams := range parameters {
+		jsonVal, err := json.Marshal(subWorkflowParams)
+		if err != nil {
+			return params, err
+		}
+		params.WorkflowParams[idx] = jsonVal
+	}
+	return params, nil
+}
+
+func EnqueueChainWorkflow(ctx context.Context, m *cereal.Manager, workflowName string, instanceName string, parameters []interface{}) error {
+	params, err := ToChainWorkflowParameters(parameters)
+	if err != nil {
+		return err
+	}
+	return m.EnqueueWorkflow(ctx, workflowName, instanceName, &params)
+}
+
+func ToChainWorkflowInstance(instance cereal.ImmutableWorkflowInstance) (*ChainWorkflowInstance, error) {
+	chainInstance := ChainWorkflowInstance{}
+	if err := instance.Err(); err != nil {
+		chainInstance.err = err
+	} else {
+		if instance.IsRunning() {
+			payload := &ChainWorkflowPayload{}
+			if err := instance.GetPayload(&payload); err != nil {
+				return nil, err
+			}
+			chainInstance.payload = payload
+		} else {
+			result := &ChainWorkflowPayload{}
+			if err := instance.GetResult(&result); err != nil {
+				return nil, err
+			}
+			chainInstance.result = result
+		}
+	}
+	chainInstance.isRunning = instance.IsRunning()
+
+	params := ChainWorkflowParams{}
+	if err := instance.GetParameters(&params); err != nil {
+		return nil, err
+	}
+	chainInstance.params = &params
+	return &chainInstance, nil
+}
+
+func GetChainWorkflowInstance(ctx context.Context, m *cereal.Manager, workflowName string, instanceName string) (*ChainWorkflowInstance, error) {
```

Comment:
I think this might be unused? Which then would make ToChainWOrkflowInstance unused as well unless it is used in one of the other PRs